### PR TITLE
refactor: Code cleanup due C4 and other audit feedback (SC-2039)

### PR DESCRIPTION
**Pull request by deluca-mike**
# Description

Rectify comments, interface and variable naming issues raised by C4 and and cleanup interfaces, state-mutability identifiers, and visibility identifiers code-wide.

Clubhouse: https://app.clubhouse.io/maplefinance/story/2039/maple-core-c4-code-cleanup

Depends on #368 

NOTE: Please ignore until above PR is merged, as this is branches off of it, and thus includes many more changes (with respect to `develop`) than expected.

# Integrations Checklist

- [X] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [X] Have any events changed or been added? If yes, outline below.
- [X] Has all documentation been updated?

# Changelog

1 - chore: Mirrored admin variables in global context, Pool, PoolFactory, Loan and LoanFactory may make it confusing for deployment and maintenance (https://github.com/code-423n4/2021-04-maple-findings/issues/36)

- Loan.admins -> Loan.loanAdmins
- Loan.setAdmin -> Loan.setLoanAdmin
- LoanFactory.admins -> LoanFactory.loanFactoryAdmins
- LoanFactory.setAdmin -> LoanFactory.setLoanFactoryAdmin
- MapleGlobals.admin -> MapleGlobals.globalAdmin
- MapleGlobals.setAdmin -> MapleGlobals.setGlobalAdmin
- Pool.admin -> Pool.poolAdmins
- Pool.setAdmin -> Pool.setPoolAdmin
- PoolFactory.admin -> PoolFactory.poolFactoryAdmins
- PoolFactory.setAdmin -> PoolFactory.setPoolFactoryAdmin

2 - refactor: Interface and implementation function declaration differs (https://github.com/code-423n4/2021-04-maple-findings/issues/98)

- Corrected non-matching interface names
- Explicit uint256 usage
- Removed unused or non-existent interfaces
- Increase interface coverage
- Remove unused variables
- Set minimal explicit state-mutabilities
- Standardized "visibility | state-mutability" pattern for functions
- Fixed one different compiler version

3 - refactor: FDT interface cleanup
- Fixed circular naming of FDT and IFDT
- Created Interfaces for all FDT variants (in their own directory)
- FDT interface inheritance all the way up to shed redundant interface declarations

4 - chore: Functions calculating the value of BPT is vulnerable to flash-loan attacks (https://github.com/code-423n4/2021-04-maple-findings/issues/113)

### Loan
- `admins` => `loanAdmins`
- `makeFullPayment() public` => `makeFullPayment() external`
- `setAdmin(address newAdmin, bool allowed)` => `setLoanAdmin(address loanAdmin, bool allowed)`

### LoanFactory
- `admins` => `loanFactoryAdmins `
- `setAdmin(address newAdmin, bool allowed)` => `setLoanFactoryAdmin(address loanFactoryAdmin, bool allowed)`

### MapleGlobals
- `admin` => `globalAdmin`
- `setAdmin(address newAdmin)` => `setGlobalAdmin(address newGlobalAdmin)`

### Pool
- `admins` => `poolAdmins `
- `setAdmin(address newAdmin, bool allowed)` => `setPoolAdmin(address poolAdmin, bool allowed)`

### PoolFactory
- `admins` => `poolFactoryAdmins `
- `setAdmin(address newAdmin, bool allowed)` => `setPoolFactoryAdmin(address poolFactoryAdmin, bool allowed)`

### LoanLib
- `getNextPayment(address superFactory, address repaymentCalc, uint256 nextPaymentDue, address lateFeeCalc)` => `getNextPayment(address repaymentCalc, uint256 nextPaymentDue, address lateFeeCalc)`

### PoolLib
- `poolSanityChecks(...) external` => `poolSanityChecks(...) external view`
- `handleDefault(IERC20  liquidityAsset, address stakeLocker, address stakeAsset,  address loan,  uint256 defaultSuffered)` => `handleDefault(IERC20  liquidityAsset, address stakeLocker, address stakeAsset,  uint256 defaultSuffered)`
- `calculateClaimAndPortions(...) external` => `calculateClaimAndPortions(...) external pure`
- `fromWad(uint256 amt, uint256 liquidityAssetDecimals) public view` => `fromWad(uint256 amt, uint256 liquidityAssetDecimals) public pure`

### ChainlinkOracle
- `getDenomination() external view` => `getDenomination() external pure`

### Loan
- `AdminSet(address admin, bool allowed)` => `LoanAdminSet(address loanAdmin, bool allowed)`

### LoanFactory
- `LoanFactoryAdminSet(address loanFactoryAdmin, bool allowed)` (new)

### MapleGlobals
- `AdminSet(address newAdmin)` => `GlobalAdminSet(address newGlobalAdmin)`

### Pool
- `AdminSet(address newAdmin, bool allowed)` => `PoolAdminSet(address poolAdmin, bool allowed)`

### PoolFactory
- `AdminSet(address newAdmin, bool allowed)` => `PoolFactoryAdminSet(address poolFactoryAdmin, bool allowed)`

### CODE REVIEW

File: contracts/library/PoolLib.sol [hunk 1]
```diff
@@ -74,7 +74,7 @@ library PoolLib {
         address dlFactory,
         uint256 amt
     ) external {
-        IGlobals globals    = _globals(superFactory);
+        IMapleGlobals globals    = _globals(superFactory);
```
Comment (lucas-manuel) [inline contracts/library/PoolLib.sol:77]: Align `=`

**Review by satyamakgec — DISMISSED** [review]

File: contracts/Pool.sol [hunk 2]
```diff
@@ -243,7 +243,7 @@ contract Pool is PoolFDT {
     */
     function _handleDefault(address loan, uint256 defaultSuffered) internal {
 
-        (uint256 bptsBurned, uint256 postBurnBptBal, uint256 liquidityAssetRecoveredFromBurn) = PoolLib.handleDefault(liquidityAsset, stakeLocker, stakeAsset, loan, defaultSuffered);
```
Comment (lucas-manuel) [inline contracts/Pool.sol:246]: Lets actually take this out and add it into the PR for this: https://app.clubhouse.io/maplefinance/story/2038/maple-core-c4-gas-optimizations

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Great PR - small changes

**Review by lucas-manuel — DISMISSED** [review]

File: contracts/library/LoanLib.sol [hunk 3]
```diff
@@ -127,8 +127,8 @@ library LoanLib {
         @param liquidityAsset Address of loan asset that is supported by the loan in other words denominated currency in which it taking funds.
         @param globals Instance of the `MapleGlobals` contract.
      */
-    function reclaimERC20(address token, address liquidityAsset, IGlobals globals) external {
-        require(msg.sender == globals.governor(),               "L:NOT_GOV");
+    function reclaimERC20(address token, address liquidityAsset, IMapleGlobals globals) external {
+        require(msg.sender == globals.governor(), "L:NOT_GOV");
         require(token != liquidityAsset && token != address(0), "L:INVALID_TOKEN");
```
Comment (lucas-manuel) [inline contracts/library/LoanLib.sol:132]: Error message alignment

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**